widgets/src/components/common/WidgetConfig.js

```diff
@@ -0,0 +1,104 @@
+import React, { useEffect, useState, useRef } from 'react';
+import PropTypes from 'prop-types';
+import Button from '@material-ui/core/Button';
+import { Close, Check } from '@material-ui/icons';
+import CircularProgress from '@material-ui/core/CircularProgress';
+
+import { getPageWidget, putPageWidget } from 'api/app-builder/pages';
+
+const WidgetConfig = ({ children, config: passedConfig, ...rest }) => {
+  const [loaded, setLoaded] = useState(false);
+  const [code, setCode] = useState({});
+  const [config, setConfig] = useState({});
+  const [saving, setSaving] = useState(false);
+  const [saved, setSaved] = useState({});
+  const configRef = useRef(null);
+
+  const { pageCode, frameId } = rest;
+
+  const handleSaveClick = async () => {
+    setSaving(true);
+
+    const configState = configRef.current.state.config;
+    const stringifiedConfig = Object.keys(configState).reduce((acc, setting) => {
+      return {
+        ...acc,
+        [setting]:
+          typeof configState[setting] === 'string'
+            ? configState[setting]
+            : JSON.stringify(configState[setting]),
+      };
+    }, {});
+    const response = await putPageWidget(
+      pageCode,
+      frameId,
+      JSON.stringify({ code, config: stringifiedConfig })
+    );
+
+    const noErrors = (response && response.errors && response.errors.length === 0) || false;
+    setSaving(false);
+    setSaved({ success: noErrors });
+    setTimeout(() => {
+      setSaved({});
+    }, 2000);
+  };
+
+  useEffect(() => {
+    async function fetchWidgetConfig() {
```

Comment:
why declaring a function here if you are calling it right after its declaration?

Reply:
Using async directly in the useEffect function isn't allowed (`useEffect function must return a cleanup function or nothing. Promises and useEffect(async () => ...) are not supported, but you can call an async function inside an effect`)